modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition_manager.py

```diff
@@ -137,6 +138,154 @@ def get_indices(cls, axis, partitions, index_func=None):
         new_idx = ray.get(new_idx)
         return new_idx[0].append(new_idx[1:]) if len(new_idx) else new_idx
 
+    @classmethod
+    def concat(cls, axis, left_parts, right_parts):
+        """
+        Concatenate the blocks of partitions with another set of blocks.
+
+        Parameters
+        ----------
+        axis : int
+            The axis to concatenate to.
+        left_parts : np.ndarray
+            NumPy array of partitions to concatenate with.
+        right_parts : np.ndarray or list
+            NumPy array of partitions to be concatenated.
+
+        Returns
+        -------
+        np.ndarray
+            A new NumPy array with concatenated partitions.
+
+        Notes
+        -----
+        Assumes that the blocks are already the same shape on the
+        dimension being concatenated. A ValueError will be thrown if this
+        condition is not met.
+        """
+        result = super(PandasOnRayDataframePartitionManager, cls).concat(
+            axis, left_parts, right_parts
+        )
+        if axis == 0:
+            return cls.rebalance_partitions(result)
+        else:
+            return result
+
+    @classmethod
+    def rebalance_partitions(cls, partitions):
+        """
+        Rebalance a 2-d array of partitions.
+
+        Rebalance the partitions by building a new array
+        of partitions out of the original ones so that:
+          - If all partitions have a length, each new partition has roughly the
```

Comment:
Not sure I fully understand this `Otherwise, each new partition spans roughly the same number of old partitions`. What is the heuristic and why do we should use it?

Reply:
By that part, I meant that we divide the existing partitions equally among all the new partitions. We do that because we don't have the length of each old partition, so we can't guarantee that we'll be able to divide the rows evenly across all the new partitions. The heuristic is that we hope each old partition has roughly the same number of rows.